### DISCUSSION TITLE
让Bot响应自己在另一端的消息

### DISCUSSION THREAD

**Discussion by mzltest** [nonebot/discussions#399] [问答] [answered]
现在是这个场景：我同一个qq有两端在线，一个是pcqq我自己用，一个是nb2挂机器人。我现在希望我在pcqq端自己发的消息也会被nb2端处理，也就是我用自己的号在qq发指令会收到我自己的响应，cq端已经打开report-self-message，但是nb2端会把我发的消息当作外发消息。

我在pcqq端发的消息在nb2输出是这样：
```
07-18 14:04:09 [SUCCESS] nonebot | OneBot V11 ****** | [message_sent]: {'time': 1689689048, 'self_id': ******, 'post_type': 'message_sent', 'group_id': xxxxxx, 'raw_message': '/ping', 'font': 0, 'anonymous': None, 'message_type': 'group', 'user_id': ******, 'message_id': -1464574434, 'message': '/ping', 'sender': {'age': 0, 'area': '', 'card': 'test', 'level': '', 'nickname': 'xxxx', 'role': 'member', 'sex': 'unknown', 'title': '', 'user_id': ******}, 'message_seq': 12698, 'sub_type': 'normal'}
```
此处******均是同一个QQ号
适配器 onebotv11。

**Comment by yanyongyu** [comment]
事件类型是 `message_sent` ，你需要 `on("message_sent")`来响应此类事件

**Reply by mzltest** [reply]
试了下，如果直接用消息很多的上下文都没有(Event has no context!)，而且貌似没法直接接入我现在用的on_command，貌似还挺麻烦，我再看看有没有其他解决方案吧，还是感谢回答。

**Comment by mnixry** [comment]
你可以通过继承Adapter，然后在事件序列化之前把`message_sent`修改成`message`，这样就能被当作一般消息事件处理。

---

@yanyongyu 要不要加个hook？感觉这个pre deserialization hook需求我遇到好几个了

**Reply by yanyongyu** [reply]
ob适配器可以添加自定义模型

**Reply by mnixry** [reply]
自定义模型和这个反序列化hook的目标需求还是不太一样的，比如题主这个需求好像自定义model就没法解决

如果没有理解错的话，他要的是把所有message_sent当作普通的message event来处理，而不是额外的自定义event

**Reply by yanyongyu** [reply]
但是修改协议数据本身就是一个hack行为，如果修改了消息类型将会导致预期外的响应行为，应该由响应器明确指出需要响应自身发送出的消息

**Reply by mnixry** [reply]
这个行为确实比较hack，但是我认为这并不能阻止我们去做这个功能

举例来说，我们有call API hook，支持替换send_msg函数，这些都会给用户带来预料外的行为

**Reply by yanyongyu** [reply]
call api hook并不能替换函数，只能替换函数的参数数据

**Reply by mzltest** [reply]
昨天看了下文档，好像看到过有一个self_id由adapter提供，我在想能不能把adapter的那个self_id改成一个不存在的固定值，这样是不是我自己发的消息也会当作收到的消息被处理，也不用动我写的代码，找的教程抄的不会改

**Comment by fu050409** [comment, marked as answer]
你可以使用`on_message_sent = on("message_sent", block=False)`，并在你的`handler`函数前装饰：
```python
@yourcommand.handle()
@on_message_sent.handle()
def yourhandler(bot: Bot, matcher: Matcher, event: Event):
    ...
```
值得注意的是，event参数应当被声明为`Event`而非`MessageEvent`或`GroupMessageEvent`，否则mathcer无法正常捕获消息。

同时，`"message_sent"`模式下的`event`无法从`get_message()`或其它方法获得消息，但你可以`event.raw_message`直接获取机器人账号的消息。你可以通过`isinstance(event, Event)`来判断当前方法是响应哪个事件来分别获得消息并响应。

**Reply by fu050409** [reply]
我不是非常确定连续两个装饰器会不会出现异常，但是我这边没出现问题。如果出现异常，可以试试将方法独立成一个函数，分别建立两个`handler`函数并用两个不同装饰器装饰，这样它们会同时指向同一个方法。

**Reply by fu050409** [reply]
顺带提一句，`event.messgae`和`event.raw_message`都是原始HTML数据，有些特殊字符，可以使用Python的`html`标准库转义特殊字符

**Reply by mzltest** [reply]
感谢解答，我这机器人这个月刚被封号（上signserver就封了)，最近应该没办法试，先给您的答案mark as answer了.

**Comment by Loping151** [comment]
两年以后我有相同需求来到了这里。我发现我的nonebot仍然无法处理message sent类型，于是我做了如下处理：

nonebot/adapters/onebot/v11/adapter.py line 441:
```
        if json_data['post_type'] == "message_sent":
            json_data["post_type"] = "message"
            json_data.pop("message_sent_type", None)
```

我自己都觉得变钛，但是目的达到了。之前使用拉格朗自己就做了这一步，可惜不维护了。

**Reply by yanyongyu** [reply]
有插件补充对应event类型，你也可以通过自己注册事件类的方式给ob添加message_sent类型事件

**Reply by Loping151** [reply]
是的，但我希望sent类型完全兼容之前所有的 is_instance(envet, PrivateMessageEvent) 之类判断，我觉得这是最简单的方法

**Reply by faithleysath** [reply]
可以使用我的fork版本，实现了这一更改。https://github.com/faithleysath/adapter-onebot
```
# nb adapter install nonebot-adapter-onebot
pip uninstall nonebot-adapter-onebot
pip install git+https://github.com/faithleysath/adapter-onebot.git
```

**Reply by Loping151** [reply]
非常好更改

我实现了中间层来统一解决这个问题
https://github.com/Loping151/BotShepherd
在其中，我的一些其他希望存在的功能也实现了

**Reply by faithleysath** [reply]
> 非常好更改
> 
> 我实现了中间层来统一解决这个问题 https://github.com/Loping151/BotShepherd 在其中，我的一些其他希望存在的功能也实现了

非常好的中间层，给你star了。
另外我注意到`nonebot2`的`o11 MessageEvent`也缺少了`target_id`这个字段，我也添加到我的fork版本里了。

**Comment by 21036045** [comment]
还有后续吗，完整教程是怎么样的

**Reply by yanyongyu** [reply]
这个问题是和你使用的适配器以及协议端强绑定的，需要知道协议端上报自己消息的类型

**Reply by 21036045** [reply]
好的，我知道了
